presto-main/src/main/java/com/facebook/presto/sql/analyzer/PostPruningTableColumnsAnalyzer.java

```diff
@@ -0,0 +1,327 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.metadata.QualifiedObjectName;
+import com.facebook.presto.sql.tree.AliasedRelation;
+import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Except;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FieldReference;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.Intersect;
+import com.facebook.presto.sql.tree.Join;
+import com.facebook.presto.sql.tree.JoinOn;
+import com.facebook.presto.sql.tree.Lateral;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.NodeRef;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.SampledRelation;
+import com.facebook.presto.sql.tree.Table;
+import com.facebook.presto.sql.tree.TableSubquery;
+import com.facebook.presto.sql.tree.Union;
+import com.facebook.presto.sql.tree.Unnest;
+import com.facebook.presto.sql.tree.Values;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ListMultimap;
+
+import java.util.List;
+
+/**
+ * Finds all fields in the query that would be read, after pruning.
+ *
+ * The high-level algorithm works as follows:
+ * 1. Find all fields referenced in all clauses of the outermost SELECT query, and add them to an explore list.
+ * 2. For each field reference F, find its relation R.
+ * 3. If R is a SELECT query:
+ *    a. Find the SELECT item expression that F references. Add all fields referenced by that expression to the explore list.
+ *    b. Add all fields referenced by every other clause of the SELECT query to the explore list.
+ * 4. Otherwise,
+ *    a. Add F's referenced field to a referenced fields list.
+ *    b. For each child of R, find the child field of F, and add it to the explore list.
+ * 5. Repeat from step 2 for all fields in the explore list, until all have been resolved to a base table relation.
+ *
+ * The pruning step takes place in step 3a, where output expressions that have not been previously referenced, are ignored (pruned).
+ * The referenced fields list at the end of this algorithm will contain all the columns referenced by the query, after pruning.
+ */
+public class PostPruningTableColumnsAnalyzer
+{
+    private final Analysis analysis;
+
+    public PostPruningTableColumnsAnalyzer(Analysis analysis)
+    {
+        this.analysis = analysis;
+    }
+
+    public void analyze(Node node)
```

Comment:
I thought we originally thought we could use the maps from Analysis for this work? That didn't work out? One issue here is if we add some new clauses you might miss them if there is no visitor.

Reply:
The visitor pattern was a lot cleaner than a graph traversal of the maps in Analysis. This would only miss something, if we add a new Relation node to the AST. Other nodes would be visited by the default traversal.